### PR TITLE
Boole domain

### PR DESCRIPTION
**Pull request by nkryuchkov**
~~The PR must be rebased onto the `boole-fork` branch after both https://github.com/ssvlabs/ssv/pull/2302 and https://github.com/ssvlabs/ssv/pull/2503 are merged~~
~~The PR must be rebased onto the `boole-fork` branch after https://github.com/ssvlabs/ssv/pull/2503 is merged~~

### CODE REVIEW

File: protocol/v2/ssv/runner/aggregator.go [hunk 1]
```diff
@@ -234,7 +234,8 @@ func (r *AggregatorRunner) ProcessConsensus(ctx context.Context, logger *zap.Log
 		Messages: []*spectypes.PartialSignatureMessage{msg},
 	}
 
-	msgID := spectypes.NewMsgID(r.BaseRunner.NetworkConfig.DomainType, r.GetShare().ValidatorPubKey[:], r.BaseRunner.RunnerRoleType)
+	domain := r.BaseRunner.NetworkConfig.DomainTypeAtSlot(decidedValue.Duty.Slot)
+	msgID := spectypes.NewMsgID(domain, r.GetShare().ValidatorPubKey[:], r.BaseRunner.RunnerRoleType)
```
Comment (y0sher) [inline protocol/v2/ssv/runner/aggregator.go:238]: should we send old aggregator runner messages using the new domain type after fork? I'm not sure about the change here..

Comment (nkryuchkov) [reply protocol/v2/ssv/runner/aggregator.go:238]: It shouldn't be possible because we won't have any aggregator runner duties during Boole epochs. We could have used `r.BaseRunner.NetworkConfig.DomainType` but I like the current code better, because it keeps the runner choice logic abstract and independent of the runner code, otherwise it would look like a hardcoded value

File: network/p2p/p2p_setup.go [hunk 2]
```diff
@@ -218,10 +218,25 @@ func (n *p2pNetwork) setupPeerServices() error {
...
+				allowedDomains = append(allowedDomains, nextDomainString)
+			}
+		}
+
+		var networkFilter connections.HandshakeFilter
+		if len(allowedDomains) == 1 {
+			networkFilter = connections.NetworkIDFilter(allowedDomains[0])
+		} else {
+			networkFilter = connections.NetworkIDFilterAny(allowedDomains...)
+		}
```
Comment (iurii-ssv) [inline network/p2p/p2p_setup.go]: Nit: I think it would be cleaner to have just 1 `NetworkIDFilterAny` func that accepts `allowedDomains...` and checks for its `len(allowedDomains)` inside of it

File: networkconfig/network.go [hunk 3]
```diff
@@ -32,6 +33,23 @@ func (n Network) StorageName() string {
...
+func (n Network) CurrentDomainType() spectypes.DomainType {
+	return n.DomainTypeAtSlot(n.EstimatedCurrentSlot())
+}
+
+func (n Network) NextDomainTypeAtSlot(slot phase0.Slot) spectypes.DomainType {
+	if n.BooleForkAtSlot(slot) {
+		return n.DomainTypeAtSlot(slot)
+	}
+	return n.NextDomainType
+}
```
Comment (iurii-ssv) [inline networkconfig/network.go]: I think `NextDomainTypeAtSlot` always returns `n.NextDomainType` (seems unit-tests don't fully cover all edge-cases) ... it should probably be something like this instead ?
```go
func (n Network) NextDomainTypeAtSlot(slot phase0.Slot) spectypes.DomainType {
	if n.BooleForkAtSlot(slot) || n.InBooleTransitionWindow(slot) {
		return n.NextDomainType
	}
	return n.DomainType
}
```

I would also add a documenting comment for "what exactly this func is for".

Also, once we are **past the Boole-fork** we'll have both `currentDomain` and `nextDomain` equal (both set to Boole) ... which is probably gonna work fine, but maybe it's better to have `nextDomain` to be explicitly set to `spectypes.DomainType{}` after transition window (and handle it explicitly wherever the handling is needed).

**Review by iurii-ssv — COMMENTED** [review]
LGTM, with maybe a minor adjustment

File: networkconfig/network.go (hunk 3 above)
Comment (nkryuchkov) [reply networkconfig/network.go]: The next domain represents the next fork domain, so it's always the same, because we don't have any next scheduled forks after Boole. I think that zeroing it out after the fork is error-prone and doesn't give us any benefits. Current domain equal to next domain after fork looks fine to me, because next domain will become redundant after fork (we already used to have next domain once, and then we removed it).

As for issues that might be caused by time drift, we allow the next domain for `InBooleTransitionWindow` during a handshake: https://github.com/ssvlabs/ssv/blob/boole-domain/network/p2p/p2p_setup.go#L224

So I'd keep it as it is now

Comment (iurii-ssv) [reply networkconfig/network.go]: > The next domain represents the next fork domain ...

Yeah, it's fine as is (I just don't like to rely on "implicit" behavior even if it works well).

But my comment is more about how this function is implemented, currently it is equivalent to just this (which seems strange to me) ?
```go
func (n Network) NextDomainTypeAtSlot(slot phase0.Slot) spectypes.DomainType {
	return n.NextDomainType
}
```

Comment (nkryuchkov) [reply networkconfig/network.go]: @iurii-ssv Yeah, I didn't think it through well enough after refactoring. Perhaps we can delete this method

Comment (nkryuchkov) [reply networkconfig/network.go]: removed

File: network/discovery/dv5_service.go [hunk 4]
```diff
@@ -235,26 +235,20 @@ func (dvs *DiscV5Service) checkPeer(ctx context.Context, e PeerEvent) error {
...
 		return false
 	}
 
-	nextDomain := dvs.nextDomainType()
-	return nodeDomainType == nextDomain || nodeNextDomainType == nextDomain
+	// During transition, accept either configured fork domain for compatibility.
+	return nodeDomainType == dvs.netCfg.DomainType ||
+		nodeNextDomainType == dvs.netCfg.DomainType ||
+		nodeDomainType == dvs.netCfg.NextDomainType ||
+		nodeNextDomainType == dvs.netCfg.NextDomainType
```
Comment (iurii-ssv) [inline network/discovery/dv5_service.go]: I wonder if we could simplify the **current**/**next** domain handling to something like this ?
- publisher will only publish the **current** domain
- the receiver would check that ^ against its **current**/**next** domain by doing something like:
```go
func (dvs *DiscV5Service) isDomainCompatible(nodeDomainType spectypes.DomainType) bool {
	if nodeDomainType == dvs.netCfg.CurrentDomainType() {
		return true
	}

	if dvs.netCfg.InBooleTransitionWindow(dvs.netCfg.EstimatedCurrentSlot()) {
		return nodeDomainType == dvs.netCfg.DomainType || nodeDomainType == dvs.netCfg.NextDomainType
	}

	return false
}
```

**Review by iurii-ssv — APPROVED** [review]
Looks good, but maybe we could [simplify the current/next domain handling a bit](https://github.com/ssvlabs/ssv/pull/2679#discussion_r2822708204)

File: network/discovery/dv5_service.go (hunk 4 above)
Comment (nkryuchkov) [reply network/discovery/dv5_service.go]: @iurii-ssv probably, it might be good to check `nodeNextDomainType == dvs.netCfg.DomainType || nodeDomainType == dvs.netCfg.NextDomainType` too to avoid any clock skew. I think it's still possible to be in the fork and get a handshake from a node that was sent before the fork

Comment (iurii-ssv) [reply network/discovery/dv5_service.go]: I'm thinking: the `InBooleTransitionWindow` is designed to take care of the clock-skews,

> I think it's still possible to be in the fork and get a handshake from a node that was sent before the fork

so once our node is past Boole-fork we'll have 1 slot of time accepting peers who broadcast the older domainType ... after that 1 slot passes those peers must switch to broadcasting the nextDomainType. It is possible that we'll ignore some peers during the transition because of p2p delays, but it doesn't seem like it should cause any real issues (while the simpler code is always preferable).

But maybe it's not worth changing it if you already have tested the current solution in this PR to work well.

Comment (iurii-ssv) [reply network/discovery/dv5_service.go]: Thinking more about it, maybe `InBooleTransitionWindow` isn't even the "right" window to use here,

what we really need is to wait a sufficient amount of time (however long it takes for all SSV nodes to start publishing/advertising themselves with the `nextDomainType` value) after Boole-fork epoch starts ... I don't think 1 slot is enough

^ we'd need that ^ if we were to simplify as I suggested above, your current solution in this PR probably works already regardless

**Review by iurii-ssv — APPROVED** [review]

File: network/discovery/dv5_service.go (hunk 4 above)
Comment (nkryuchkov) [reply network/discovery/dv5_service.go]: Well, actually, we QA'd the version you suggested. I decided to extend it, but thinking more about it, it's likely redundant. I simplified it

File: operator/validator/controller.go [hunk 5]
```diff
@@ -1043,8 +1043,6 @@ func SetupCommitteeRunners(
 ) validator.CommitteeRunnerFunc {
 	buildController := func(role spectypes.RunnerRole) *qbftcontroller.Controller {
 		config := &qbft.Config{
-			BeaconSigner: options.Signer,
-			Domain:       options.NetworkConfig.DomainType,
```
Comment (olegshmuelov) [inline operator/validator/controller.go:1047]: Now that `BeaconSigner` and `Domain` are no longer set, the fields, their getters (`GetShareSigner`, `GetSignatureDomainType`), and the `signing` interface in `config.go` are dead code

File: network/discovery/dv5_service.go [hunk 6]
```diff
@@ -233,6 +235,18 @@ func (dvs *DiscV5Service) checkPeer(ctx context.Context, e PeerEvent) error {
...
 }
 
+func (dvs *DiscV5Service) isDomainCompatible(nodeDomainType spectypes.DomainType, nodeNextDomainType spectypes.DomainType) bool {
+	if nodeDomainType == dvs.netCfg.CurrentDomainType() {
+		return true
+	}
+
+	if !dvs.netCfg.InBooleTransitionWindow(dvs.netCfg.EstimatedCurrentSlot()) {
+		return false
+	}
```
Comment (olegshmuelov) [inline network/discovery/dv5_service.go:245]: After the 1-slot subsequent window, peers whose cached ENR still shows `DomainType=Alan` are rejected even if `NextDomainType=Boole` signals fork compatibility.

`PublishENR` fires at fork time via subnet changes, but discv5 ENR propagation isn't instant - peers discovered before the fork can have stale cached ENRs beyond the 1-slot window.

Consider also accepting `nodeNextDomainType == CurrentDomainType()` outside the transition window:

```go
if nodeDomainType == dvs.netCfg.CurrentDomainType() || nodeNextDomainType == dvs.netCfg.CurrentDomainType() {
    return true
}
```

File: message/validation/signed_ssv_message.go [hunk 7]
```diff
@@ -121,15 +119,6 @@ func (mv *messageValidator) validateSSVMessage(ssvMessage *spectypes.SSVMessage)
 		return e
 	}
 
-	// Rule: If domain is different then self domain
```
Comment (olegshmuelov) [inline message/validation/signed_ssv_message.go:124]: Removing this lets messages with random domains reach the per-`MessageID` validation locks and committee lookups before `validateDomainAtSlot` rejects them - potential amplification vector since each random domain creates a unique lock entry (TTL = 2 epochs).

Consider a cheap slot-free early gate:

```go
domain := ssvMessage.GetID().GetDomain()
if !bytes.Equal(domain, mv.netCfg.DomainType[:]) && !bytes.Equal(domain, mv.netCfg.NextDomainType[:]) {
    return ErrWrongDomain
}
```

**Review by olegshmuelov — APPROVED** [review]

**Comment by nkryuchkov** [comment]
The current state of the branch has been passed to the QA team, as this branch is the only thing left to make the `boole-fork` branch work overall.

Any further changes to this PR are unwanted unless they are critical

**Review by y0sher — APPROVED** [review]
